server/src/handler_driver.c

```diff
@@ -168,6 +192,412 @@ static bool drive_route_params_valid(const char *train_id, t_interlocking_route
+		syslog_server(LOG_ERR, 
+		              "Get multi segment indices for route %s: Unable to allocate memory", route->id);
+		return d_seg_indices;
+	}
+	// For every segment in route->path, check if it occurs again at a different position. 
+	// If yes, add both occurrences to d_seg_indices.
+	for (size_t path_index_i = 0; path_index_i < route->path->len; ++path_index_i) {
+		const char *path_item_i = g_array_index(route->path, char *, path_index_i);
+		if (path_item_i != NULL && is_type_segment(path_item_i)) {
+			for (size_t path_index_n = 0; path_index_n < route->path->len; ++path_index_n) {
+				const char *path_item_n = g_array_index(route->path, char *, path_index_n);
+				if (path_item_n != NULL && is_type_segment(path_item_n)
+				        && strcmp(path_item_i, path_item_n) == 0 && path_index_i != path_index_n) {
+					d_seg_indices.arr[d_seg_indices.len] = path_index_n;
+					d_seg_indices.len++;
+				}
+			}
+		}
+	}
+	if (d_seg_indices.len == 0) {
+		free(d_seg_indices.arr);
+		d_seg_indices.arr = NULL;
+	} else {
+		size_t *tmp = reallocarray(d_seg_indices.arr, d_seg_indices.len, sizeof(size_t));
+		if (tmp == NULL) {
+			syslog_server(LOG_ERR, 
+			              "Get multi segment indices for route %s: Unable to re-allocate memory", 
+			              route->id);
+			free(d_seg_indices.arr);
+			d_seg_indices.arr = NULL;
+			d_seg_indices.len = 0;
+		} else {
+			// re-alloc successful
+			d_seg_indices.arr = tmp;
+		}
+	}
+	return d_seg_indices;
+}
+
+// Queries the position of the train. Checks which segments that the train occupies exist in the route path. 
+// Return -3 on param err, -2 on train not on tracks, -1 on train not on route, >= 0 if train is on route.
+// If train occupies at least one segment in route path, returns the route-path index of the occupied segment
+// that is the furthest along the route. Ignores any segments that appear more than once in route->path
+ssize_t get_train_pos_index_in_route_no_multis(const char* train_id, const t_interlocking_route *route, 
+                                                 const t_route_multi_segment_indices *multi_segment_indices) {
+	if (train_id == NULL || route == NULL || route->path == NULL || multi_segment_indices == NULL) {
+		return -3;
+	}
+	t_bidib_train_position_query train_pos_query = bidib_get_train_position(train_id);
+	if (train_pos_query.segments == NULL || train_pos_query.length == 0) {
+		bidib_free_train_position_query(train_pos_query);
+		return -2;
+	}
+	ssize_t train_pos_index = -1;
+	size_t train_pos_index_unsigned = 0;
+	const size_t path_count = route->path->len;
+	for (size_t i = 0; i < train_pos_query.length; ++i) {
+		// Search starting at most recent pos_index to skip unnecessary comparisons
+		// If index is present in array of multi-occurring segments, skip iteration 
+		for (size_t n = train_pos_index_unsigned; n < path_count; ++n) {
+			bool ignore = false;
+			for (size_t k = 0; k < multi_segment_indices->len; ++k) {
```

Comment:
It would be more efficient if `multi_segment_indices` mapped each `path_index` to `true` (segment is repeated) or `false` (segment is not repeated). This would avoid the costly searches through `multi_segment_indices`.

Example: `bool ignore = multi_segment_indices[n];`

Reply:
Pretty good idea. I originally thought that the count of repeatedly occurring segments is not astronomously big, meaning that searching often would be not too much of a performance hit. The direct mapping is much more efficient though, will change it in the second adjustment commit to separate this from the other changes made.